src/charm.py

```diff
@@ -7,226 +7,91 @@
 """Charm the service."""
 
 import logging
-import socket
-from typing import Dict, List, Optional
-
-from charms.loki_k8s.v1.loki_push_api import _PebbleLogClient
-from cosl import JujuTopology
-from loki import (
-    CONTAINER_NAME,
-    LOKI_CERT_FILE,
-    Loki,
-)
-from loki_cluster import (
-    ConfigReceivedEvent,
-    LokiClusterRequirer,
-    LokiRole,
-)
-from ops.charm import CharmBase, CollectStatusEvent
-from ops.framework import BoundEvent, Object
+import re
+from typing import Optional
+
+from cosl.coordinated_workers.worker import CERT_FILE, CONFIG_FILE, Worker
+from ops.charm import CharmBase
 from ops.main import main
-from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import Layer
 
 # Log messages can be retrieved using juju debug-log
 logger = logging.getLogger(__name__)
 
+CONTAINER_NAME = "loki"
+LOKI_PORT = 3100
+
 
 class LokiWorkerK8SOperatorCharm(CharmBase):
     """A Juju Charmed Operator for Loki."""
 
     def __init__(self, *args):
         super().__init__(*args)
-        self._container = self.unit.get_container(CONTAINER_NAME)
-        self._loki = Loki(self._container, self._loki_roles, self.model)
-        self._topology = JujuTopology.from_charm(self)
-        self._loki_cluster = LokiClusterRequirer(self)
-        self._log_forwarder = ManualLogForwarder(
+        self._worker = Worker(
             charm=self,
-            loki_endpoints=self._loki_cluster.get_loki_endpoints(),
+            name="loki",
+            pebble_layer=self.pebble_layer,
+            endpoints={"cluster": "loki-cluster", "tracing": "tracing"},
```

Comment:
This likely needs to be changed, if you pull the latest `cos-lib` you'll see the "tracing" key isn't there anymore